File: hellojava/src/main/java/com/fathzer/chess/optimized/IntMove.java
package com.fathzer.chess.optimized;

/**
 * Utility class for encoding and decoding move information into a single int.
 * The encoding uses the following bit layout:
 * - from (7 bits): bits 0-6 (0-127)
 * - to (7 bits): bits 7-13 (0-127)
 * - capture (5 bits): bits 14-18 (0-31)
 */
public final class IntMove {
    private static final int FROM_MASK = 0x7F;        // 7 bits for from (0-127)
    private static final int TO_MASK = 0x3F80;         // 7 bits for to (shifted left by 7)
    private static final int CAPTURE_MASK = 0x7C000;    // 5 bits for capture (shifted left by 14)
    
    private static final int TO_SHIFT = 7;
    private static final int CAPTURE_SHIFT = 14;

    private IntMove() {
        // Private constructor to prevent instantiation
    }

    public static int fromUCI(Board board, String uci) {
        if (uci.length() != 4) {
            throw new IllegalArgumentException("Invalid UCI move: " + uci);
        }
        final int toSquare = Board.getSquare(uci.substring(2));
        final int capturedPiece = board.getPiece(toSquare);
        return move(Board.getSquare(uci.substring(0, 2)), toSquare, capturedPiece);
    }

    
    /**
     * Creates a packed move integer from individual components.
     * @param from The source square (0-127)
     * @param to The target square (0-127)
     * @param capture The capture value (0-31)
     * @return A packed integer containing all three values
     */
    public static int move(int from, int to, int capture) {
        return (from) | 
               (to << TO_SHIFT) | 
               (capture << CAPTURE_SHIFT);
    }
    
    /**
     * Extracts the source square from a packed move.
     * 
     * @param move The packed move
     * @return The source square (0-127)
     */
    public static int from(int move) {
        return move & FROM_MASK;
    }
    
    /**
     * Extracts the target square from a packed move.
     * 
     * @param move The packed move
     * @return The target square (0-127)
     */
    public static int to(int move) {
        return (move & TO_MASK) >>> TO_SHIFT;
    }
    
    /**
     * Extracts the capture value from a packed move.
     * 
     * @param move The packed move
     * @return The capture value (0-31)
     */
    public static int capture(int move) {
        return (move & CAPTURE_MASK) >>> CAPTURE_SHIFT;
    }
    
    public static String toString(int move) {
    	final int capture = capture(move);
        if (capture==0) {
            return Board.getUCI(from(move)) + Board.getUCI(to(move));
        } else {
            return Board.getUCI(from(move)) + "x" + Board.getUCI(to(move)) + "(" + Piece.getCode(capture) + ")";
        }
    }
}

File: hellogo/pkg/chess/optimized/intmove.go
// Package optimized provides optimized data structures for chess engine implementation.
package optimized

// IntMove provides utility functions for encoding and decoding move information into a single int.
// The encoding uses the following bit layout:
// - from (7 bits): bits 0-6 (0-127)
// - to (7 bits): bits 7-13 (0-127)
// - capture (5 bits): bits 14-18 (0-31)
const (
	fromMask    = 0x7F    // 7 bits for from (0-127)
	toMask      = 0x3F80  // 7 bits for to (shifted left by 7)
	captureMask = 0x7C000 // 5 bits for capture (shifted left by 14)

	toShift      = 7
	captureShift = 14
)

// FromUCI creates a packed move from a UCI string and the current board state.
// The UCI string should be 4 characters long (e.g., "e2e4").
func FromUCI(board *Board, uci string) int {
	if len(uci) != 4 {
		panic("Invalid UCI move: " + uci)
	}
	toSquare := GetSquare(uci[2:])
	capturedPiece := board.getPiece(toSquare)
	return Move(GetSquare(uci[:2]), toSquare, capturedPiece)
}

// Move creates a packed move integer from individual components.
// from: The source square (0-127)
// to: The target square (0-127)
// capture: The capture value (0-31)
// Returns a packed integer containing all three values.
func Move(from, to, capture int) int {
	return (from) |
		(to << toShift) |
		(capture << captureShift)
}

// From extracts the source square from a packed move.
// Returns the source square (0-127).
func From(move int) int {
	return move & fromMask
}

// To extracts the target square from a packed move.
// Returns the target square (0-127).
func To(move int) int {
	return (move & toMask) >> toShift
}

// Capture extracts the capture value from a packed move.
// Returns the capture value (0-31).
func Capture(move int) int {
	return (move & captureMask) >> captureShift
}

// String returns a string representation of the move in UCI format.
// For captures, it also includes the captured piece in parentheses.
func String(move int) string {
	capture := Capture(move)
	if capture == 0 {
		return GetUCI(From(move)) + GetUCI(To(move))
	}
	return GetUCI(From(move)) + "x" + GetUCI(To(move)) + "(" + string(GetCode(capture)) + ")"
}
